cl/api/utils.py

```diff
@@ -346,6 +349,148 @@ def allow_request(self, request, view):
+        self.history = self.cache.get(self.key, [])
+        self.request_timestamp = self.timer()
+
+        # Drop any requests from the history which have now passed the
+        # throttle duration
+        while self.history and self.history[-1][-1] <= self.request_timestamp:
+            self.history.pop()
+
+        citations_in_history = sum(
+            citation_count for citation_count, timestamps in self.history
+        )
+
+        if citations_in_history >= max_num_citations:
+            self.throttle_request(request)
+
+        self.save_citation_count(request, view)
+
+    def save_citation_count(self, request, view):
+        """
+        Inserts the number of citations and the expiration time along with
+        the key into the cache.
+        """
+        citation_count = self.get_citation_count_from_request(request, view)
+        if not citation_count:
+            return
+
+        max_num_citations, duration = self.get_citations_rate(request)
+        expiration = (
+            citation_count * (max_num_citations / duration)
+            if citation_count > max_num_citations
+            else duration
+        )
+        self.history.insert(
+            0,
+            [
+                citation_count,
+                self.request_timestamp + expiration,
+            ],
+        )
+
+        self.cache.set(self.key, self.history, expiration)
+
+    def allow_request(self, request, view):
+        self.throttle_request_by_citation_count(request, view)
+        return super().allow_request(request, view)
+
+    def throttle_request(self, request):
+        """
+        This helper iterates through the request history in reverse
+        chronological order to calculate the soonest time a new request can be
+        made and raises the `Throttled` exception with the details.
+
+        The exception includes details about the throttling:
+
+        - `error_message`: A message indicating the request was throttled.
+        - `wait_until`: An ISO 8601 formatted string representing the soonest
+                    time the next request can be made without throttling.
+
+        Args:
+            request: The request object to be throttled.
+
+        Raises:
+            Throttled: The exception includes details about the throttling.
+        """
+        max_num_citations, _ = self.get_citations_rate(request)
+        soonest_time = None
+        for idx in reversed(range(len(self.history))):
+            remaining_citation = sum(
+                citation_count for citation_count, _ in self.history[:idx]
+            )
+            if remaining_citation < max_num_citations or not idx:
+                datetime_obj = datetime.fromtimestamp(
+                    self.history[idx][-1], timezone.utc
+                )
+                soonest_time = datetime_obj.isoformat()
+
+        raise Throttled(
+            detail={
+                "error_message": "Request was throttled.",
```

Comment:
We'd have to update line 474 to change `_` to `unit`, but if we did, we could do:

```suggestion
                "error_message": f"Too many requests (allowed rate: {max_num_citations}/{unit}).",
```

Reply:
Thanks for the suggestion, I'll tweak the `get_citations_rate` method so we can include the rate in the error message